src/main/java/bio/terra/stairway/Flight.java

```diff
@@ -0,0 +1,177 @@
+ *
+ *  ISSUE: handling InterruptedException - is there anything that the Flight level should do
+ *  for handling this?
+ */
+public class Flight implements Callable<FlightResult> {
+    static class StepRetry {
+        private Step step;
+        private RetryRule retryRule;
+
+        StepRetry(Step step, RetryRule retryRule) {
+            this.step = step;
+            this.retryRule = retryRule;
+        }
+    }
+
+    private List<StepRetry> steps;
+    private Database database;
+    private FlightContext flightContext;
+
+    public Flight(SafeHashMap inputParameters) {
+        flightContext = new FlightContext(inputParameters).flightClassName(this.getClass().getName());
+        steps = new LinkedList<>();
+    }
+
+    public FlightContext context() {
+        return flightContext;
+    }
+
+    void setFlightContext(FlightContext flightContext) {
+        this.flightContext = flightContext;
+    }
+
+    void setDatabase(Database database) {
+        this.database = database;
+    }
+
+    // Used by subclasses to build the step list with default no-retry rule
+    protected void addStep(Step step) {
+        steps.add(new StepRetry(step, RetryRuleNone.getRetryRuleNone()));
+    }
+
+    // Used by subclasses to build the step list with a retry rule
+    protected void addStep(Step step, RetryRule retryRule) {
+        steps.add(new StepRetry(step, retryRule));
+    }
+
+    /**
+     * Call may be called for a flight that has been interrupted and is being recovered
+     * so we may be headed either direction.
+     */
+    public FlightResult call() {
+        try {
+            if (context().isDoing()) {
+                StepResult doResult = runSteps();
+                if (doResult.isSuccess()) {
+                    return new FlightResult(doResult, context().getWorkingMap());
+                }
+
+                // Remember the failure from the do; that is what we want to return
+                context().result(doResult);
+                context().doing(false);
+
+                // Record the step failure and direction change in the database
+                database.step(context());
+            }
+
+            StepResult undoResult = runSteps();
+            if (undoResult.isSuccess()) {
+                // Return the error from the doResult - that is why we failed
+                return new FlightResult(context().getResult(), context().getWorkingMap());
+            }
+
+            // Record the undo failure
+            database.step(context());
+
+            // Dismal failure - undo failed!
+            return FlightResult.flightResultFatal(
+                    new StairwayExecutionException("Dismal failure: " + undoResult.getThrowable().toString(),
+                            context().getResult().getThrowable().orElse(null)));
+        } catch (InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            return FlightResult.flightResultFatal(ex);
+        }
+    }
```

Comment:
I'm in favor of trying to minimize the number of return statements in a method. Maybe this can be broken up into smaller methods?

Reply:
I prefer avoiding deeply nested logic by using return statements. Breaking it into smaller methods works for me...

Reply:
I looked at trying to re-vamp this code. I couldn't figure out a good way to do it. Instead, I added a bunch of comments that provide more explanation.